Add .travis.yml

isaacbrodsky opened the pull request:
Add Travis CI to the repo. `.travis.yml` is added with three configuration (Linux GCC, Linux Clang, and one for testing formatting.) Added the Apache license and Travis build icons to the readme.

A few build issues are fixed in utilities and tests. A few possible crashes are fixed in test code.

Commit:
Add .travis.yml

nrabinowitz reviewed (approved):
Looks good, thanks for the test cleanup.

ibgreen reviewed (approved):
LGTM

Commit:
Change tests to not leak allocations.

The tests sometimes did not free allocated structures because they would simply be free'd when the test case exits.
